types/material-ui/index.d.ts

```diff
@@ -712,16 +712,16 @@ declare namespace __MaterialUI {
         type?: string | undefined;
         containerElement?: React.ReactNode | string | undefined;
     }
-
-    export interface EnhancedButtonProps extends React.HTMLAttributes<{}>, SharedEnhancedButtonProps<EnhancedButton> {
+    type Union<T, S> = Omit<T, keyof S> & S;
```

Comment:
Could you clarify what this type is for? There's no union operator used here but intersection operators. Why the omit?

Reply:
Agree about the naming (it's a union of properties). The Omit is there to omit properties that would clash:

```ts
interface SharedProps { className?: string | undefined; }
interface React.HTMLAttributes<T> { 
  className?: string | undefined; 
  title?: string | undefined; 
}
interface MyComponentProps extends SharedProps,  React.HTMLAttributes<{}> {}
```

This will work, but after changing HTMLAttributes to `className?: string | undefined | null` you get

```ts
interface SharedProps { className?: string | undefined; }
interface React.HTMLAttributes<T> { 
  className?: string | undefined | null; 
  title?: string | undefined | null; 
}
interface MyComponentProps extends SharedProps,  React.HTMLAttributes<{}> {}
```

which doesn't because SharedProps doesn't accept null (assuming your TS config has strictNullChecks enabled). 

To avoid this I chose to Omit from HTMLAttributes all the fields defined in SharedProps. This would cause the type of properties defined in SharedProps to remain the same as before this change.

```ts
type MyComponentProps = SharedProps & Omit<HTMLAttributes<{}>, keyof SharedProps>;
// = SharedProps & { title?: string | undefined | null;  } 
// = { 
//  className?: string | undefined; 
//  title?: string | undefined | null;  
// }
```